### PR TITLE
prov/gni: perf improvements

### PR DESCRIPTION
**Pull request by hppritcha**
Try to recycle elements off free lists more frequently
to reduce cache misses.

Reduce poll rate for cm nic unless there are elements
on the cm nic work queue.

This shaves about 50-100 nsecs off of the osu_latency results:

before at 37c2af66

```
# OSU MPI Latency Test v5.0
# Size          Latency (us)
0                       2.06
1                       2.04
2                       2.03
4                       2.03
8                       2.04
16                      2.03
32                      2.04
64                      2.04
128                     2.03
256                     2.04
512                     2.07
1024                    2.39
2048                    2.71
4096                    3.37
8192                    4.98
16384                  10.75
32768                  11.87
65536                  15.76
131072                 22.05
262144                 35.01
524288                 61.87
1048576               114.77
2097152               239.83
4194304               515.08
```

with optimizations

```
# OSU MPI Latency Test v5.0
# Size          Latency (us)
0                       2.01
1                       1.99
2                       1.98
4                       1.97
8                       1.98
16                      1.98
32                      1.98
64                      1.99
128                     1.97
256                     1.98
512                     2.00
1024                    2.27
2048                    2.61
4096                    3.30
8192                    4.82
16384                  10.62
32768                  11.72
65536                  15.46
131072                 21.67
262144                 34.64
524288                 61.23
1048576               114.16
2097152               240.17
4194304               512.34
```

using

```
srun -n 2 -N 2 --exclusive --cpu_bind=socket --nodelist=nid00051,nid00050
```

command line options

@e-harvey 
@ztiffany 

Signed-off-by: Howard Pritchard howardp@lanl.gov

### CODE REVIEW

**Comment by hppritcha** [comment]
cori is offline today so LANL jenkins will not run.

**Comment by ztiffany** [comment]
:+1:

File: prov/gni/src/gnix_cm_nic.c [hunk 1]
```diff
@@ -236,13 +237,17 @@ int _gnix_cm_nic_progress(struct gnix_cm_nic *cm_nic)
...
-		ret = _gnix_dgram_poll(cm_nic->dgram_hndl,
-					  GNIX_DGRAM_NOBLOCK);
-		if (ret != FI_SUCCESS) {
-			GNIX_WARN(FI_LOG_EP_CTRL,
-				"_gnix_dgram_poll returned %s\n",
-				  fi_strerror(-ret));
-				goto err;
+		++cm_nic->poll_cnt;
+		if (((cm_nic->poll_cnt % 512) == 0)  ||
+			!dlist_empty(&cm_nic->cm_nic_wq)) {
```
Comment (e-harvey) [inline prov/gni/src/gnix_cm_nic.c:242]: If the cm_nic->cm_nic_wq is empty more often than the cm_nic->poll_cnt reaches 512, would it be faster to short circuit this if statement by moving !dlist_empty(&cm_nic->cm_nic_wq) to the left of the logical or?

File: prov/gni/src/gnix_cq.c [hunk 2]
```diff
@@ -249,8 +251,6 @@ static struct slist_entry *alloc_cq_entry(size_t size)
 	return NULL;
 }
 
-extern int _gnix_cm_nic_progress(struct gnix_cm_nic *cm_nic);
```
Comment (e-harvey) [inline prov/gni/src/gnix_cq.c:252]: Is this prototype no longer needed?

**Comment by e-harvey** [comment]
:+1:

File: prov/gni/src/gnix_cq.c (hunk 2 above)
Comment (hppritcha) [inline prov/gni/src/gnix_cq.c:252]: yep because I added gnix_cm_nic.h include at the top of the file.

File: prov/gni/src/gnix_cm_nic.c (hunk 1 above)
Comment (hppritcha) [inline prov/gni/src/gnix_cm_nic.c:242]: i don't think so because its an ||.  compiler has to check both conditionals.

**Pull request merged**